com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderPropertyCollection.cs

```diff
@@ -0,0 +1,57 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using BlockProperty = UnityEditor.ShaderFoundry.BlockVariable;
+
+namespace UnityEditor.ShaderFoundry
+{
+    internal class ShaderPropertyCollection
+    {
+        Dictionary<string, BlockProperty> visitedProperties = new Dictionary<string, BlockProperty>();
+        List<BlockProperty> shaderProperties = new List<BlockProperty>();
+
+        public bool ReadOnly { get; private set; }
+        public IEnumerable<BlockProperty> Properties => shaderProperties;
+
+        public void Add(BlockProperty property)
+        {
+            if (ReadOnly)
+            {
+                Debug.LogError("ERROR: attempting to add property to readonly collection");
+                return;
+            }
+
+            if (visitedProperties.TryGetValue(property.Name, out var existingProperty))
+            {
+                ValidateAreEquivalent(property, existingProperty);
+                return;
+            }
+
+            visitedProperties.Add(property.Name, property);
+            shaderProperties.Add(property);
+        }
+
+        public void AddRange(IEnumerable<BlockProperty> properties)
+        {
+            foreach (var property in properties)
```

Comment:
I would suggest to add a check for ReadOnly here as well and early out - otherwise it will spam this error for each property.

Reply:
This is really an authoring problem, not a content problem, so this should never happen in normal flow. I'll add the early out error though for cleanliness.